include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,1085 @@
+
+  /**
+   * Set entity index at which quadrature points are accessed. This can, e.g.,
+   * a cell index, a cell batch index, or a face batch index.
+   */
+  template <typename T = FERemoteEvaluationCommunicatorType>
+  typename std::enable_if<
+    std::is_same<typename T::FERemoteEvaluationDataViewType,
+                 internal::FERemoteEvaluationDataView<false>>::value,
+    void>::type
+  reinit(const unsigned int index)
+  {
+    data_offset = comm->get_shift(index);
+  }
+
+  /**
+   * Set cell, face pair at which quadrature points are accessed.
+   */
+  template <typename T = FERemoteEvaluationCommunicatorType>
+  typename std::enable_if<
+    std::is_same<typename T::FERemoteEvaluationDataViewType,
+                 internal::FERemoteEvaluationDataView<true>>::value,
+    void>::type
+  reinit(const unsigned int cell_index, const unsigned int face_number)
+  {
+    data_offset = comm->get_shift(cell_index, face_number);
+  }
+
+  /**
+   * Get the value at quadrature point @p q. The entity on which the values
+   * are defined is set via `reinit()`.
+   */
+  const value_type
+  get_value(const unsigned int q) const
+  {
+    Assert(data_offset != numbers::invalid_unsigned_int,
+           ExcMessage("reinit() not called."));
+    AssertIndexRange(data_offset + q, data.values.size());
+    return data.values[data_offset + q];
+  }
+
+  /**
+   * Get the gradients at quadrature pointt @p q. The entity on which the
+   * gradients are defined is set via `reinit()`.
+   */
+  const gradient_type
+  get_gradient(const unsigned int q) const
+  {
+    Assert(data_offset != numbers::invalid_unsigned_int,
+           ExcMessage("reinit() not called."));
+    AssertIndexRange(data_offset + q, data.gradients.size());
+    return data.gradients[data_offset + q];
+  }
+
+private:
+  /**
+   * Use Triangulation as MeshType.
+   */
+  void
+  set_mesh(const Triangulation<dimension> &tria)
+  {
+    this->tria = &tria;
+  }
+
+  /**
+   * Use DoFHandler as MeshType.
+   */
+  void
+  set_mesh(const DoFHandler<dimension> &dof_handler)
+  {
+    this->dof_handler = &dof_handler;
+  }
+
+  /**
+   * Data that is accessed by `get_value()` and `get_gradient()`.
+   */
+  FERemoteEvaluationDataType data;
+
+  /**
+   * Underlying communicator which handles update of the ghost values and
+   * gives position of values and gradients stored in
+   * FERemoteEvaluationData.
+   */
+  SmartPointer<const FERemoteEvaluationCommunicatorType> comm;
+
+  /**
+   * Used MeshType.
+   */
+  SmartPointer<const Triangulation<dimension>> tria;
+  SmartPointer<const DoFHandler<dimension>>    dof_handler;
+
+  /**
+   * Flags that indicate which ghost values are updated.
+   */
+  const VectorTools::EvaluationFlags::EvaluationFlags vt_flags;
+
+  /**
+   * First selected component.
+   */
+  const unsigned int first_selected_component = 0;
+
+  /**
+   * Offset to data after last call of `reinit()`.
+   */
+  unsigned int data_offset = numbers::invalid_unsigned_int;
```

Comment:
This as well.

Reply:
Should `start` and `cell_start` in `FEREDataView` also be initialized in a constructor to conform with the `deal.II` style (even though no constructor is present yet) or is it OK?

See: https://github.com/dealii/dealii/pull/15982/files#diff-c360108d85e19ac86c9dd9a82f59988c6611b77a137adeae4b9a0442052e4dbaR307